_drafts/test-automation-renode.md

```diff
@@ -0,0 +1,886 @@
+
+It does this using the monitor commands `Save` and `Load`. Wouldn't it be cool
+if we could call the `Save` command in CI for failing tests and then locally run
+`Load` in Renode to get the exact state of the device at the time of failure?
+Yes it would be.
+
+Renode has part of this
+[built-in](https://github.com/renode/renode/blob/05377ef375daa3d5ea0de12633d27bf26e20b3b3/src/Renode/RobotFrameworkEngine/renode-keywords.robot#L77-L86).
+Unfortunately, this is yet another piece of functionality that doesn't exist in
+the current public release, so I've
+[copied the code](https://github.com/memfault/interrupt-renode-test-automation/blob/master/tests/common.robot)
+(and overrode the snapshot save location) so that these snapshots get uploaded
+into our Github Actions artifacts!
+
+I've download an example ZIP archive from one of the builds from GitHub and
+extracted it.
+
+```
+$ tree .
+test_results
+├── log.html
+├── logs.txt
+├── report.html
+├── robot_output.xml
+├── snapshots
+│   └── test-basic-Trigger_Fault.fail.save
+└── test-basic.xml
+```
+
+The `test-basic-Trigger_Fault.fail.save` file is a snapshot from a test and
+shell command that forces a crash.
+
+```c
+int cli_command_fault(int argc, char *argv[]) {
+    void (*g_bad_func_call)(void) = (void (*)(void))0x20000002;
+    g_bad_func_call();
+    return 0;
+}
+```
+
+Let's load it up into Renode & GDB and see what it looks like. I've created
+another script to help with loading these save files into Renode,
+`load-save.sh`.
+
+```
+$ ./load-save.sh test_results/snapshots/test-basic-Trigger_Fault.fail.save
+Loaded monitor commands from: ./scripts/monitor.py
+Monitor available in telnet mode on port 1234
+machine-0: GDB server with all CPUs started on port :3333
+```
+
+We see that the GDB port on 3333 is up and running, so let's attach to that
+port. Don't forget to include the ELF file (which is also saved in the GitHub
+build artifacts).
+
+```
+$ arm-none-eabi-gdb-py --eval-command="target remote :3333" --se renode-example.elf
+[...]
+(gdb) bt
+#0  blocking_handler () at ../../cm3/vector.c:104
+#1  <signal handler called>
+#2  0x20000002 in rcc_ahb_frequency ()
+#3  0x08000252 in cli_command_fault (argc=1, argv=0x2000ff74) at src/app_shell_commands.c:24
+#4  0x08000528 in prv_process () at src/shell/src/shell.c:113
+#5  0x080005ec in shell_receive_char (c=13 '\r') at src/shell/src/shell.c:147
+#6  0x080001e0 in main () at src/app.c:29
+```
+
+That looks right! `0x20000002` was the bogus address I used which caused a
+UsageFault, and we can tell it came from the CLI command `fault`.
+
+## Improving The Renode Tooling
+
+To improve my workflow with Renode, I needed to do something about the
+Mono-emulated terminal windows that were spawned with Renode. Although not
+entirely documented, it's quite easy to never have to use these terminal windows
+directly and you can instead attach to Renode using Telnet. This enables you to
+have use iTerm2, your native clipboard, tmux, etc.
```

Comment:
Wait! You forgot the most important part: how do I do this? What's this `start-headless.sh` script you mention? I want to run Renode headless, found your post, but didn't find the answer :'(